resolwe/flow/models/history_manager.py

```diff
@@ -0,0 +1,142 @@
+        self._trackers: dict[
+            tuple[Type[models.Model], str], Type["TrackChange"]
+        ] = dict()
+        self._tracked_fields: dict[Type[models.Model], list[str]] = defaultdict(list)
+
+    def add_tracker(self, tracker: Type["TrackChange"]):
+        """Add tracker to the manager.
+
+        :raises AssertionError: when tracker for the same field already exists.
+        """
+        for model, name in tracker.applicable_to:
+            assert (
+                model,
+                name,
+            ) not in self._trackers, (
+                f"Tracker for field '{name}' and model '{model}' already exists."
+            )
+            self._trackers[(model, name)] = tracker
+            self._tracked_fields[model].append(name)
+
+    def valid_value(
+        self,
+        instance: "HistoryMixin",
+        field_name: str,
+        timestamp: Optional[datetime] = None,
+    ) -> Any:
+        """Get the valid value at the point in time.
+
+        :attr timestamp: a point in time. When not provided the current moment is used.
+        :raises AttributeError: when no value exists for the given timestamp.
+        """
+        # Take the latest change before the timestamp. When timestamp is not provided
+        # take the latest change.
+        history = instance.get_history_object()
+        tracker = self._trackers[(instance._meta.model, field_name)]
+        if timestamp:
+            tracker_qset = tracker.objects.filter(
+                history=history, timestamp__lte=timestamp
+            )
+        else:
+            tracker_qset = tracker.objects.filter(history=history)
+
+        change = tracker_qset.last()
+        if change is None:
+            raise RuntimeError(
+                f"No value exists for history object {history.pk}, field "
+                f"{tracker.applicable_to} and timestamp {timestamp}."
+            )
+        return change.get_value(field_name)
+
+    def process_change(self, instance: "HistoryMixin"):
+        """Process the change in the field.
+
+        :raises KeyError: when tracker for the field is not registered.
+        :raises AssertionError: when the instance does not have history tracking.
+        """
+        model = instance._meta.model
+        for field_name in self._tracked_fields[model]:
+            if (model, field_name) not in self._trackers:
+                continue
+            try:
+                tracker = self._trackers[(model, field_name)]
+                if hasattr(instance, field_name):
+                    current_value = self.valid_value(instance, field_name)
+                    new_value = getattr(instance, field_name)
+                    if current_value != new_value:
+                        tracker.process_change(instance, field_name)
+            except RuntimeError:
+                tracker.process_change(instance, field_name)
+
+
+history_manager = HistoryManager()
+
+
+def post_delete_signal(sender, instance, **kwargs):
```

Comment:
Move this to the `delete` method.

Reply:
I was just about to do that when a paragraph

```The delete() method does a bulk delete and does not call any delete() methods on your models. It does, however, emit the [pre_delete](https://docs.djangoproject.com/en/4.2/ref/signals/#django.db.models.signals.pre_delete) and [post_delete](https://docs.djangoproject.com/en/4.2/ref/signals/#django.db.models.signals.post_delete) signals for all deleted objects (including cascaded deletions).```

from Django documentation changed my mind. However, I do not like to depend on signals to much so I will move this to a `delete` method. And keep in mind that bulk delete is not a good idea.